### PR TITLE
Wrap route handler and dispatch `failed` event

### PR DESCRIPTION
**Pull request by AlmightyYakob**
This PR adds a `failed` rest event, in the form of `rest.method.group.failed`. This will fire if an exception is raised during the request, re-raising the exception as to not modify behavior.

### CODE REVIEW

File: girder/api/rest.py [hunk 1]
```diff
@@ -965,7 +970,14 @@ def handleRoute(self, method, path, params):
...
             self._defaultAccess(handler)
-            val = handler(**kwargs)
+
+            # Attempt to handle and trigger failed event if an
+            # exception is raised.
+            try:
+                val = handler(**kwargs)
+            except Exception as e:
+                events.trigger('.'.join((eventPrefix, 'failed')), kwargs)
+                raise e
```
Comment (zachmullen) [inline girder/api/rest.py:980]:
```suggestion
            except Exception:
                events.trigger('.'.join((eventPrefix, 'failed')), {'kwargs': kwargs})
                raise
```

**Review by zachmullen — CHANGES_REQUESTED** [review]

**Comment by manthey** [comment]
We should add this event to the docs in docs/plugin-development.rst.

File: girder/api/rest.py (hunk 1 above)
Comment (AlmightyYakob) [reply girder/api/rest.py:980]: I didn't know you could re-raise like that, thanks!

What's the reason for wrapping `kwargs` in a dict? It seems to differ from how the `before` and `after` events include these.

Comment (zachmullen) [reply girder/api/rest.py:980]: I don't have strong feelings, but passing it wrapped in an outer dict allows easily adding more info the event in the future, without breaking compatibility for anyone.

Comment (AlmightyYakob) [reply girder/api/rest.py:980]: I suppose that's true (I could image passing the exception itself as well) but I'd think consistency with the before and after events would be more important. Maybe not, @manthey any thoughts?

Comment (manthey) [reply girder/api/rest.py:980]: I'd rather have the consistency.  To my knowledge, we haven't wanted to extend the before and after events.

**Review by zachmullen — APPROVED** [review]

**Comment by manthey** [comment]
For some reason, the ansible test is brittle and can fail without indicating an actual problem.  This can be merged.

**Review by manthey — APPROVED** [review]

**Comment by AlmightyYakob** [comment]
@manthey @zachmullen I'm not authorized to merge so one of you will need to actually hit merge.

**Pull request merged**